File: medium/462. Minimum Moves to Equal Array Elements II.js
// 462. Minimum Moves to Equal Array Elements II

// Given an integer array nums of size n, return the minimum number of moves required to make all array elements equal.

// In one move, you can increment or decrement an element of the array by 1.

// Test cases are designed so that the answer will fit in a 32-bit integer.

 

// Example 1:

// Input: nums = [1,2,3]
// Output: 2
// Explanation:
// Only two moves are needed (remember each move increments or decrements one element):
// [1,2,3]  =>  [2,2,3]  =>  [2,2,2]
// Example 2:

// Input: nums = [1,10,2,9]
// Output: 16
 

// Constraints:

// n == nums.length
// 1 <= nums.length <= 105
// -109 <= nums[i] <= 109


// #1
const minMoves2 = function(nums) {
  nums.sort((a, b) => a - b)
  
  const median = Math.floor(nums.length / 2)
  let count = 0
  
  for (let num of nums) {
    count += Math.abs(nums[median] - num)
  }
  
  return count
};

// #2
const minMoves22 = function(nums) {
  nums.sort((a,b) => a - b);
  let start = 0;
  let end = nums.length - 1;
  let moves = 0;

  while (start <= end) {
    moves += nums[end] - nums[start];
    start++;
    end--;
  }

  return moves;
};

File: medium/462. Minimum Moves to Equal Array Elements II.py
# 462. Minimum Moves to Equal Array Elements II

# 1
class Solution:
  def minMoves2(self, nums: List[int]) -> int:
    sorted_nums = sorted(nums)
    med = sorted_nums[int(len(sorted_nums) / 2)]
    res = 0
    for n in sorted_nums: 
      res += abs(n - med)
    return res


# 2
class Solution:
  def minMoves2(self, nums: List[int]) -> int:
    sorted_nums = sorted(nums)
    res = 0
    l = 0
    r = len(nums) - 1
    while l <= r: 
      res += sorted_nums[r] - sorted_nums[l]
      l += 1
      r -= 1
    return res